sklearn/_engine/base.py

```diff
@@ -0,0 +1,187 @@
+            engine = getattr(engine, attr)
+        return engine
+
+
+def _parse_entry_point(entry_point):
+    module_name, engine_qualname = entry_point.value.split(":")
+    provider_name = next(iter(module_name.split(".", 1)))
+    return EngineSpec(entry_point.name, provider_name, module_name, engine_qualname)
+
+
+@lru_cache
+def _parse_entry_points(provider_names=None):
+    specs = []
+    all_entry_points = entry_points()
+    if hasattr(all_entry_points, "select"):
+        engine_entry_points = all_entry_points.select(group=SKLEARN_ENGINES_ENTRY_POINT)
+    else:
+        engine_entry_points = all_entry_points.get(SKLEARN_ENGINES_ENTRY_POINT, ())
+    for entry_point in engine_entry_points:
+        try:
+            spec = _parse_entry_point(entry_point)
+            if provider_names is not None and spec.provider_name not in provider_names:
+                # Skip entry points that do not match the requested provider names.
+                continue
+            specs.append(spec)
+        except Exception as e:
+            # Do not raise an exception in case an invalid package has been
+            # installed in the same Python env as scikit-learn: just warn and
+            # skip.
+            warnings.warn(
+                f"Invalid {SKLEARN_ENGINES_ENTRY_POINT} entry point"
+                f" {entry_point.name} with value {entry_point.value}: {e}"
+            )
+    if provider_names is not None:
+        observed_provider_names = {spec.provider_name for spec in specs}
+        missing_providers = set(provider_names) - observed_provider_names
+        if missing_providers:
+            raise RuntimeError(
+                "Could not find any provider for the"
+                f" {SKLEARN_ENGINES_ENTRY_POINT} entry point with name(s):"
+                f" {', '.join(repr(p) for p in sorted(missing_providers))}"
+            )
+    return specs
+
+
+def list_engine_provider_names():
+    """Find the list of sklearn_engine provider names
+
+    This function only inspects the metadata and should trigger any module import.
+    """
+    return sorted({spec.provider_name for spec in _parse_entry_points()})
+
+
+def _get_engine_classes(engine_name, provider_names, engine_specs, default):
+    specs_by_provider = {}
+    for spec in engine_specs:
+        if spec.name != engine_name:
+            continue
+        specs_by_provider.setdefault(spec.provider_name, spec)
+
+    for provider_name in provider_names:
+        if inspect.isclass(provider_name):
+            # The provider name is actually a ready-to-go engine class.
+            # Instead of a made up string to name this ad-hoc provider
+            # we use the class itself. This mirrors what the user used
+            # when they set the config (ad-hoc class or string naming
+            # a provider).
+            engine_class = provider_name
+            if getattr(engine_class, "engine_name", None) != engine_name:
```

Comment:
I think all classes that want to be an engine should have a `engine_name` attribute. Otherwise it is unclear what engine the class implements. Imagine you wrap a pipeline that has several steps that each have a plugin.

So along those lines, maybe we should use `if engine_class.engine_name != engine_name:` here, which would lead to an exception if the class doesn't define a `engine_name`.

Reply:
+1 letting it fail if the attribute is not defined

The default cython engine for kmeans doesn't define this attribute. Even if it's useless here, it would be helpful to define it for the same reasons `accepts` is ?

Reply:
> The default cython engine for kmeans doesn't define this attribute.

That is true. I forgot that for the default engine and all engines that aren't "ad-hoc" engines (is this a good name for classes directly passed in, compared to those registered via an entrypoint?) the look up has an additional level of indirection via a engine spec. The engine spec has a `name` attribute that contains the name of the engine. Just `name` seemed to generic as a name of the attribute for the "ad-hoc" classes, but maybe we should unify it?

I'm unsure about defining it also for the non ad-hoc engine classes. It adds an additional location where this information is stored (attribute of the class and name of the entrypoint) and they could become out of sync.

Reply:
Maybe the engine name could only be defined on the class itself and we could update the engine specs in the importlib metadata to only register the class?

But that means that we will need to import the class to discover its name, which is a bit sad, because it could trigger import overhead from plugins unrelated to the specific algorithm of interest.

Reply:
But maybe this is not a big deal.

Reply:
We could add a `engine_name` attribute to the engine class when it is loaded via `EngineSpec.get_engine_class`?